File: ai_interview/api/parser.py
# /api/parser.py
import re
import pdfplumber
import phonenumbers
from typing import Dict, Optional

class ResumeParser:
    def __init__(self):
        # ... (Your entire ResumeParser class code goes here) ...
        # Common name patterns (this is a simplified approach)
        self.name_patterns = [
            r'^[A-Z][a-z]+ [A-Z][a-z]+$',
            r'^[A-Z][a-z]+ [A-Z]\. [A-Z][a-z]+$',
        ]
    
    def extract_text_from_pdf(self, pdf_file) -> str:
        text = ""
        try:
            with pdfplumber.open(pdf_file) as pdf:
                for page in pdf.pages:
                    text += page.extract_text() + "\n"
        except Exception as e:
            print(f"Error extracting text from PDF: {e}")
        return text
    
    def extract_email(self, text: str) -> Optional[str]:
        email_pattern = r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'
        emails = re.findall(email_pattern, text)
        return emails[0] if emails else None
    
    def extract_phone_number(self, text: str) -> Optional[str]:
        try:
            for match in phonenumbers.PhoneNumberMatcher(text, "US"):
                # Return the first valid number found
                return phonenumbers.format_number(
                    match.number, phonenumbers.PhoneNumberFormat.E164
                )
        except Exception as e:
            print(f"Phone number extraction error: {e}")
        return None
    
    def extract_name(self, text: str) -> Optional[str]:
        lines = text.split('\n')
        name_indicators = [
            'summary', 'objective', 'experience', 'education', 
            'skills', 'projects', 'contact'
        ]
        
        for line in lines[:10]: # Check first 10 lines
            line = line.strip()
            if not line:
                continue
            
            words = line.split()
            if 2 <= len(words) <= 4: # Allow for middle names/initials
                if all(word.istitle() or (len(word) == 2 and word[1] == '.') for word in words):
                    if not any(indicator in line.lower() for indicator in name_indicators):
                        return line
        return None
    
    def parse_resume(self, pdf_file) -> Dict[str, Optional[str]]:
        text = self.extract_text_from_pdf(pdf_file)
        
        if not text:
            return {'error': 'Could not extract text from PDF'}
        
        name = self.extract_name(text)
        email = self.extract_email(text)
        phone = self.extract_phone_number(text)
        
        return {
            'name': name,
            'email': email,
            'phone': phone
        }

File: ai_interview/api/utils/parser.js
import pdf from "pdf-parse";
// import { PhoneNumberUtil, PhoneNumberFormat } from "google-libphonenumber";
import pkg from "google-libphonenumber";

const { PhoneNumberUtil, PhoneNumberFormat } = pkg;

const phoneUtil = PhoneNumberUtil.getInstance();

function extractName(text) {
  const lines = text.split("\n").slice(0, 10);
  const indicators = ["summary", "objective", "experience", "education", "skills", "projects", "contact"];

  for (let line of lines) {
    line = line.trim();
    if (!line) continue;

    const words = line.split(/\s+/);
    if (words.length >= 2 && words.length <= 4) {
      const isTitleCase = words.every(
        (w) => /^[A-Z][a-z]+$/.test(w) || /^[A-Z]\.$/.test(w)
      );
      if (isTitleCase && !indicators.some((i) => line.toLowerCase().includes(i))) {
        return line;
      }
    }
  }
  return null;
}

function extractEmail(text) {
  const match = text.match(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/);
  return match ? match[0] : null;
}

function extractPhone(text) {
  try {
    const match = text.match(/(\+?\d[\d\s().-]{7,}\d)/);
    if (match) {
      const number = phoneUtil.parseAndKeepRawInput(match[0], "US");
      return phoneUtil.format(number, PhoneNumberFormat.E164);
    }
  } catch (e) {
    console.error("Phone parse error:", e);
  }
  return null;
}

export async function parseResume(buffer) {
  try {
    const data = await pdf(buffer);
    const text = data.text;

    if (!text) {
      return { error: "Could not extract text from PDF" };
    }

    const name = extractName(text);
    const email = extractEmail(text);
    const phone = extractPhone(text);

    return { name, email, phone };
  } catch (e) {
    console.error("Parse error:", e);
    return { error: "Failed to parse PDF" };
  }
}
